### PR TITLE
Added KafkaClusterRebalance CRD and related tests

### PR DESCRIPTION
**Pull request by ppatierno**
Signed-off-by: Paolo Patierno <ppatierno@live.com>

### Type of change

- Enhancement / new feature

### Description

This PR is about adding the KafkaClusterRebalance CRD used by the user for asking for a rebalance dry-run first and then executing the actual proposal.

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Update/write design documentation in `./design`
- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md

### CODE REVIEW

File: api/src/test/resources/io/strimzi/api/kafka/model/KafkaClusterRebalance-with-goals-verbose.yaml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+apiVersion: kafka.strimzi.io/v1alpha1
+kind: KafkaClusterRebalance
+metadata:
+  name: my-rebalance
+spec:
+  goals:
+    - DiskCapacityGoal
```
Comment (tomncooper) [inline api/src/test/resources/io/strimzi/api/kafka/model/KafkaClusterRebalance-with-goals-verbose.yaml]: So these goals have to be pre-installed in CC. Do we know what happens if CC gets a bad/not-installed goal name? Does it throw an error? Should we proporgate that back to the user in the CR status?

Comment (tomncooper) [reply api/src/test/resources/io/strimzi/api/kafka/model/KafkaClusterRebalance-with-goals-verbose.yaml]: This is probably a job for later but I thought I should flag it

Comment (ppatierno) [reply api/src/test/resources/io/strimzi/api/kafka/model/KafkaClusterRebalance-with-goals-verbose.yaml]: it was asked by @tombentley as well on the other PR ... let me report back here my answer.

You can find the list of goals here https://github.com/linkedin/cruise-control#goals. It's possible for a CC user writing its own goal library but it's something that we are not going to support right now. It should be something possible like we do for Kafka Connect: the user writes the goal, build a jar and should include it into a custom Docker image to use as image for the cruise control. Tbh I am fine on listing some of the goals in the description but I would not add any constraint on the CRD itself. If the user set not existing goals in the list, the operator will send the request to CC anyway but will get an error which will be reported in the status of the custom resource.

File: api/src/main/java/io/strimzi/api/kafka/model/KafkaClusterRebalanceSpec.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        return goals;
+    }
+
+    public void setGoals(List<String> goals) {
+        this.goals = goals;
+    }
+
+    @Description("Enable the verbose mode for the JSON string describing the optimization result in the related status")
+    @JsonInclude(JsonInclude.Include.NON_NULL)
+    public boolean isVerbose() {
```
Comment (scholzj) [inline api/src/main/java/io/strimzi/api/kafka/model/KafkaClusterRebalanceSpec.java]: How big or coplicate is the verbose and non-verbose output? Just curious how much we need it.

File: api/src/main/java/io/strimzi/api/kafka/model/KafkaClusterRebalanceSpec.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    private static final long serialVersionUID = 1L;
+
+    private List<String> goals;
+    private boolean verbose = false;
+
+    private Map<String, Object> additionalProperties = new HashMap<>(0);
+
+    @Description("A list of goals to use in generating the proposal and executing it")
+    public List<String> getGoals() {
+        return goals;
```
Comment (scholzj) [inline api/src/main/java/io/strimzi/api/kafka/model/KafkaClusterRebalanceSpec.java]: How are the goals defined?

File: install/cluster-operator/048-Crd-kafkaclusterrebalance.yaml [hunk 4]
```diff
@@ -0,0 +1,57 @@
+apiVersion: apiextensions.k8s.io/v1beta1
```
Comment (scholzj) [inline install/cluster-operator/048-Crd-kafkaclusterrebalance.yaml]: It looks like this is missing in the helm chart?

File: api/pom.xml [hunk 5]
```diff
@@ -142,6 +142,7 @@
                                 <argument>io.strimzi.api.kafka.model.KafkaMirrorMaker=${pom.basedir}${file.separator}..${file.separator}install${file.separator}cluster-operator${file.separator}045-Crd-kafkamirrormaker.yaml</argument>
                                 <argument>io.strimzi.api.kafka.model.KafkaBridge=${pom.basedir}${file.separator}..${file.separator}install${file.separator}cluster-operator${file.separator}046-Crd-kafkabridge.yaml</argument>
                                 <argument>io.strimzi.api.kafka.model.KafkaConnector=${pom.basedir}${file.separator}..${file.separator}install${file.separator}cluster-operator${file.separator}047-Crd-kafkaconnector.yaml</argument>
+                                <argument>io.strimzi.api.kafka.model.KafkaClusterRebalance=${pom.basedir}${file.separator}..${file.separator}install${file.separator}cluster-operator${file.separator}048-Crd-kafkaclusterrebalance.yaml</argument>
```
Comment (scholzj) [inline api/pom.xml]: You need to add it also to the section for generating docs!

File: api/src/main/java/io/strimzi/api/kafka/model/KafkaClusterRebalanceSpec.java (hunk 3 above)
Comment (tombentley) [reply api/src/main/java/io/strimzi/api/kafka/model/KafkaClusterRebalanceSpec.java]: @scholzj see conversation [here](https://github.com/strimzi/strimzi-kafka-operator/pull/2306/files#r356703289).

@ppatierno I think we need a better `@Description` here.

Comment (ppatierno) [reply api/src/main/java/io/strimzi/api/kafka/model/KafkaClusterRebalanceSpec.java]: @tombentley what do you have in mind? Listing all the possible goals already available on CC (they are almost 20 or so).

Comment (tombentley) [reply api/src/main/java/io/strimzi/api/kafka/model/KafkaClusterRebalanceSpec.java]: Well, you could mention a few choice goals and/or link to some documentation which discusses goals in more detail (that could be the CC upstream doc for now, and maybe some Strimzi docs in the future).

**Comment by ppatierno** [comment]
@tombentley @scholzj can you have another pass please?

File: api/src/main/java/io/strimzi/api/kafka/model/KafkaClusterRebalanceSpec.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    private static final long serialVersionUID = 1L;
+
+    private List<String> goals;
+    private boolean verbose = false;
+
+    private Map<String, Object> additionalProperties = new HashMap<>(0);
+
+    @Description("A list of goals to use in generating the proposal and executing it. " +
+            "The supported goals are available at https://github.com/linkedin/cruise-control#goals." +
+            "Providing empty goals means to use all of them.")
```
Comment (tombentley) [inline api/src/main/java/io/strimzi/api/kafka/model/KafkaClusterRebalanceSpec.java]: > Providing empty goals means to use all of them.

All of them in what order (I assume order matters?)

Comment (ppatierno) [reply api/src/main/java/io/strimzi/api/kafka/model/KafkaClusterRebalanceSpec.java]: as far as I understood, the goals are in decreasing priority order as they are provided to the CC REST API as parameter. If you don't provide goals (so empty as I mentioned in the description), then CC uses the goals specified in the `default.goals` parameter in the configuration properties file. In that case, it's the same, the priority is the order as they are listed there.
How does it sound to you ...

"A list of goals, ordered by decreasing priority, to use in generating the proposal and executing it. " +
"The supported goals are available at https://github.com/linkedin/cruise-control#goals." +
"Providing empty goals means to use all the ones declared in the default.goals configuration parameter"

Comment (tombentley) [reply api/src/main/java/io/strimzi/api/kafka/model/KafkaClusterRebalanceSpec.java]: How about "Providing empty goals means to use all the ones declared in the default.goals cruise control configuration parameter" (to make it clear where that default.goals is)?

Comment (ppatierno) [reply api/src/main/java/io/strimzi/api/kafka/model/KafkaClusterRebalanceSpec.java]: I am fine with it.

**Comment by ppatierno** [comment]
@tombentley any more comments on this?

**Pull request merged**